File: 39_combinationSum.cpp
// Time Complexity : O(N)
// Space Complexity :O(N)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
public:
    vector<vector<int>> result;
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        if(candidates.empty() || candidates.size() == 0)
            return result;
        vector<int> path;
        recurse(candidates, target, 0, path);
        return result;
    }
    void recurse(vector<int>& candidates, int target, int index, vector<int> path){
        //base
        if(target == 0){
            result.push_back(path);
            return;
        }
        if(target<0 || index == candidates.size())
            return;
        //logic
        //do not choose
        recurse(candidates, target, index+1, path);
        path.push_back(candidates[index]);
        //choose
        recurse(candidates, target - candidates[index], index, path);
    }
};

//choosing first and than not choosing
class Solution {
public:
    vector<vector<int>> result;
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        if(candidates.empty() || candidates.size() == 0)
            return result;
        vector<int> path;
        recurse(candidates, target, 0, path);
        return result;
    }
    void recurse(vector<int>& candidates, int target, int index, vector<int> path){
        //base
        if(target == 0){
            result.push_back(path);
            return;
        }
        if(target<0 || index == candidates.size())
            return;
        //logic
        
        //choose
        vector<int> copy(path);
        copy.push_back(candidates[index]);
        recurse(candidates, target - candidates[index], index, copy);
        //do not choose
        recurse(candidates, target, index+1, path);
    }
};

//backtracking
class Solution {
public:
    vector<vector<int>> result;
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        if(candidates.empty() || candidates.size() == 0)
            return result;
        vector<int> path;
        recurse(candidates, target, 0, path);
        return result;
    }
    void recurse(vector<int>& candidates, int target, int index, vector<int> path){
        //base
        if(target == 0){
            result.push_back(path);
            return;
        }
        if(target<0 || index == candidates.size())
            return;
        //logic
        //do not choose
        
        recurse(candidates, target, index+1, path);
        //choose
        //action
        path.push_back(candidates[index]);
        recurse(candidates, target - candidates[index], index, path);
        //backtract
        path.pop_back();
    }
};

//using loop and recursion
class Solution {
public:
    vector<vector<int>> result;
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        if(candidates.empty() || candidates.size() == 0)
            return result;
        vector<int> path;
        recurse(candidates, target, 0, path);
        return result;
    }
    void recurse(vector<int>& candidates, int target, int index, vector<int> path){
        //base
        if(target == 0){
            result.push_back(path);
            return;
        }
        if(target<0)
            return;
        //logic
        for(int i=index; i<candidates.size(); i++){
            vector<int> copy(path);
            copy.push_back(candidates[i]); // used copy becuase we are choosing first so.
            recurse(candidates, target-candidates[i], i, copy);
        }
    }
};

//backtrack and loop
class Solution {
public:
    vector<vector<int>> result;
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        if(candidates.empty() || candidates.size() == 0)
            return result;
        vector<int> path;
        recurse(candidates, target, 0, path);
        return result;
    }
    void recurse(vector<int>& candidates, int target, int index, vector<int> path){
        //base
        if(target == 0){
            result.push_back(path);
            return;
        }
        if(target<0)
            return;
        //logic
        for(int i=index; i<candidates.size(); i++){
            //action
            path.push_back(candidates[i]);
            //recursion
            recurse(candidates, target-candidates[i], i, path);
            //backtrack
            path.pop_back();
        }
    }
};

File: 39-combinationSum.java
//Approach-1: Exhaustive Search
//Overall Time Complexity:O(2^(M*N))
//Overall Space Compelxity:O(2^(M*N))
class Solution {
    List<List<Integer>> result;
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        result = new ArrayList<>();
        if(candidates == null || candidates.length == 0) return result;
        dfs(candidates, 0, target, new ArrayList<Integer>());
        return result;
    }
    
    public void dfs(int[] candidates, int indx, int target, List<Integer> path){
        //Base Case
        if(target == 0){
            result.add(path);
            return;
        }
        if(target < 0 || indx == candidates.length) return;
            
        //Functionality
        //Not-Select Case
        dfs(candidates, indx + 1, target, new ArrayList<>(path));
        
        //Select-case
        path.add(candidates[indx]);
        dfs(candidates, indx, target - candidates[indx], new ArrayList<>(path));
    }
}




//Approach-2: Back-tracking
//Overall Time Complexity:O(2^(M*N))
//Overall Space Compelxity:O(2^(M*N))
class Solution {
    List<List<Integer>> result;
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        result = new ArrayList<>();
        if(candidates == null || candidates.length == 0) return result;
        dfs(candidates, 0, target, new ArrayList<Integer>());
        return result;
    }
    
    public void dfs(int[] candidates, int indx, int target, List<Integer> path){
        //Base Case
        if(target == 0){
            result.add(new ArrayList<>(path));
            return;
        }
        if(target < 0 || indx == candidates.length) return;
            
        //Functionality
        //Not-Select Case
        dfs(candidates, indx + 1, target, path);
        
        //Select-case
        path.add(candidates[indx]);
        dfs(candidates, indx, target - candidates[indx], path);
        path.remove(path.size() - 1);
    }
}



//Approach-3: ForLoop Based Recursion - Maintaining a pivot
//Overall Time Complexity:O(2^(M*N))
//Overall Space Compelxity:O(2^(M*N))
class Solution {
    List<List<Integer>> result;
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        result = new ArrayList<>();
        if(candidates == null || candidates.length == 0) return result;
        dfs(candidates, 0, target, new ArrayList<Integer>());
        return result;
    }
    
    public void dfs(int[] candidates, int pivot, int target, List<Integer> path){
        //Base Case
        if(target == 0){
            result.add(new ArrayList<>(path));
            return;
        }
        if(target < 0) return;
        
        for(int i = pivot; i < candidates.length; i++){
            List<Integer> temp = new ArrayList<>(path);
            temp.add(candidates[i]);
            dfs(candidates, i, target - candidates[i], temp);
        }
    }
}



//Approach-4: ForLoop Based Recursion - Maintaining a pivot (Backtracking)
//Overall Time Complexity:O(2^(M*N))
//Overall Space Compelxity:O(2^(M*N))
class Solution {
    List<List<Integer>> result;
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        result = new ArrayList<>();
        if(candidates == null || candidates.length == 0) return result;
        dfs(candidates, 0, target, new ArrayList<Integer>());
        return result;
    }
    
    public void dfs(int[] candidates, int pivot, int target, List<Integer> path){
        //Base Case
        if(target == 0){
            result.add(new ArrayList<>(path));
            return;
        }
        if(target < 0) return;
        
        for(int i = pivot; i < candidates.length; i++){
            path.add(candidates[i]);
            dfs(candidates, i, target - candidates[i], path);
            path.remove(path.size() - 1);
        }
    }
}



